### DISCUSSION TITLE
[HW2] Please provide the code with more context

### DISCUSSION THREAD

**Discussion by yuneg11** [seominjoon/kaist-ai605#95] [Q&A] [answered]
The provided skeleton code is as follows.

```python
class SimilaritySearch:
    def __init__(self):
        raise NotImplementedError()

    def train(self, documents: list):
        raise NotImplementedError()

    # Add documents (a list of text)
    def add(self, documents: list):
        raise NotImplementedError()

    # Returns the indices of top-k documents among the added documents
    # that are most similar to the input query 
    def search(self, query: str, k: int) -> list:
        raise NotImplementedError()
```

I'm not sure why both `train` and `add` methods require `documents` as an argument. When I first saw the code, I think the `add` method can be called multiple times to add documents to the target set, and the `train` method is called to make an index(vectors) of the target set. However, I realized that the `train` method also gets the documents.

Can you please clarify the purpose of `documents` in the `train` method?

**Comment by seominjoon** [comment, marked as answer]
I apologize for the delayed answer. Apparently the notification got messed up after changing my phone :( 

Regarding your question, training and adding are two different things.
For TF-IDF, you train to create a vocab and also compute  IDF values of the words in the vocab. 
So it is possible to train just on the subset of documents (if the corpus is very large), but you still need to add all the documents.

**Reply by emjay73** [reply]
I'm not able to understand the following sentence.
"So it is possible to train just on the subset of documents (if the corpus is very large)"
May I ask for a more detailed explanation?

**Reply by seominjoon** [reply]
What I mean is that, `train` and `add` have different purposes. For TF-IDF, `train` is to create the vocab and determine the IDF value of each term. `add` is for actually adding the documents to the index so that they are searchable. These documents are often the same, but they don't have to be. For instance, `train` can take in only a subset of documents you want to `add`.

**Reply by emjay73** [reply]
I thought input to the train() function is the 'SQUAD train set' and input to the add() is the 'SQUAD validation set'.
But if it is not the case and the train() function can take in only a subset of documents we want to add, is this statement has the same meaning that we can put 'SQUAD validation set' both to the train() function and the add() function?

**Reply by seominjoon** [reply]
Oh I see where the confusion is coming from. I talked about 'subset'  just for an example, but not applicable here because the corpus is not large. In this case, you just want to use validation data for both `train` and `add` for the best performance. It's not cheating to use it in `train` because you are only using the documents and not the questions.

**Reply by emjay73** [reply]
I see. 
So just for the cross-check, is it true that all I need in this assignment is a validation set?

**Reply by seominjoon** [reply]
Yes you are right. In general, for TF-IDF, you don't really need to have a separate training data. You can just use the data that you will use for validation or test.

**Reply by emjay73** [reply]
Thank you!
